### DISCUSSION TITLE
Training text detection model with a custom dataset - target ValueError

### DISCUSSION THREAD

**Discussion by lkosh** [mindee/doctr#1714] [Q&A] [answered]
Hello, I'm trying to train a text detection model on a custom dataset. I have formatted the input data according to https://github.com/mindee/doctr/tree/main/references/detection, here's how my labels.json looks like:
{'895f2a02-1f6a-4cf1-9c6b-dfef97b842d9_0.jpeg': {'img_dimensions': (2353,
   1653),
  'img_hash': 'a6780a176ae94e43a69eca09e1673e5e12351f4fdc11fb83c09b1a47d7d362f6',
  'polygons': [[[116, 1873], [220, 1873], [220, 1891], [116, 1891]],
   [[223, 1873], [289, 1873], [289, 1891], [223, 1891]],
   [[290, 1873], [365, 1873], [365, 1891], [290, 1891]],
   [[366, 1873], [394, 1873], [394, 1889], [366, 1889]],
   [[398, 1873], [448, 1873], [448, 1889], [398, 1889]],
   [[451, 1873], [521, 1873], [521, 1890], [451, 1890]],
...

Here, img_dimensions is a tuple (height, width).

During training, after many iterations, I encounter the following error:
File "/root/doctr/doctr/models/detection/differentiable_binarization/pytorch.py", line 252, in compute_loss                 [0/1702]
[rank1]:     targets = self.build_target(target, out_map.shape[1:], False)  # type: ignore[arg-type]                     
[rank1]:   File "/root/doctr/doctr/models/detection/differentiable_binarization/base.py", line 288, in build_target                            
[rank1]:     raise ValueError("the 'boxes' entry of the target is expected to take values between 0 & 1.")               
[rank1]: ValueError: the 'boxes' entry of the target is expected to take values between 0 & 1. 

Additionally, I notice that in Resize transform module, "if np.max(target) <= 1" condition is true for most samples, but false for a few. 

I would be very greatful if someone could explain the origin of this error and suggest how to modify the training data in order to resolve this issue.

**Comment by felixdittrich92** [comment, marked as answer]
Hi @lkosh :wave:,

Looks like an small overflow:

modify: https://github.com/mindee/doctr/blob/9045dcfc9c5c837b06fcda8e802f7cf1d95bd18c/doctr/transforms/modules/pytorch.py#L80

to:

```
if target is not None:
     target = np.clip(target, 0, 1)
```

I will open a PR to fix this soon.

Best regards,
Felix

**Reply by lkosh** [reply]
Hi Felix, Thank you for a quick response and the bugfix. I would also like to know if this is error indicates that my training data is bad/inconsistent. The reason I'm asking is that the detection model doesn't go above 55 mean IoU on eval dataset, which suggests that I'm either not using the right input data/training args, or that my dataset contains bad bboxes or rotated documents. I would be very greatful for any advice or tips on how to debug this. Thank you !

**Reply by felixdittrich92** [reply]
Mh the bug is more a problem with a small overflow in numpy (which can happen sometimes - at boxes which are really close to the document edges) - opened already a PR to fix this : #1715 

Depends a bit: 
- how many samples are in your dataset (train/val) 
- the complexity of your dataset (documents) 
- the args you used (pretrained / freeze_backbone / learning rate / scheduler / etc.)

In docTR it's a design decision to keep the training open in scripts and don't wrap it in any "Trainer" class - so feel free to modify the script to your own needs.

Maybe you can use `optuna` or something else to do a grid search to find the best params for your dataset for furthewr improvements :)

Btw.: You should focus more on precision and recall  - mIoU is really hard to reach (in our pre-training runs on ~250K real samples we reach also only something ~70% mIoU)
